include/engine/geospatial_query.hpp

```diff
@@ -208,16 +211,50 @@ template <typename RTreeT, typename DataFacadeT> class GeospatialQuery
     // Returns the nearest phantom node. If this phantom node is not from a big component
     // a second phantom node is return that is the nearest coordinate in a big component.
     std::pair<PhantomNode, PhantomNode>
-    NearestPhantomNodeWithAlternativeFromBigComponent(const util::Coordinate input_coordinate) const
+    NearestPhantomNodeWithAlternativeFromBigComponent(const util::Coordinate input_coordinate,
+                                                      const engine::StartSide start_side) const
     {
         bool has_small_component = false;
         bool has_big_component = false;
         auto results = rtree.Nearest(
             input_coordinate,
-            [this, &has_big_component, &has_small_component](const CandidateSegment &segment) {
+            [this, &start_side, &input_coordinate, &has_big_component, &has_small_component](
+                const CandidateSegment &segment) {
                 auto use_segment = (!has_small_component ||
                                     (!has_big_component && !segment.data.component.is_tiny));
                 auto use_directions = std::make_pair(use_segment, use_segment);
+                bool isOnewaySegment = !(segment.data.forward_segment_id.enabled &&
+                                         segment.data.reverse_segment_id.enabled);
+                if (!isOnewaySegment && start_side != BOTH)
+                {
+                    // Check the counter clockwise
+                    // segment.data.u ---------------- segment.data.v
+                    //                       |
+                    //                       |
+                    //                       |
+                    //                       |
+                    //                       |
+                    //                       |
+                    //                       |
+                    //                 input_coordinate
+
+                    bool input_coordinate_is_at_right = !util::coordinate_calculation::isCCW(
```

Comment:
`util::coordinate_calculation::isCCW(coordinates[segment.data.u], input_coordinate, coordinates[segment.data.v]);` to correspond the figure above